DigitalLearningSolutions.Web.Tests/Controllers/Login/LoginControllerTests.cs

```diff
@@ -163,5 +165,94 @@ public void Log_in_with_approved_delegate_id_fetches_associated_admin_user()
             A.CallTo(() => loginService.GetVerifiedAdminUserAssociatedWithDelegateUser(testDelegate, A<string>._))
                 .MustHaveHappened();
         }
+
+        [Test]
+        public void Multiple_available_centres_should_redirect_to_ChooseACentre_page()
+        {
+            // Given
+            var expectedAdminUser = UserTestHelper.GetDefaultAdminUser(centreId: 1, centreName: "Centre 1");
+            var expectedDelegateUsers = new List<DelegateUser>
+                { UserTestHelper.GetDefaultDelegateUser(centreId: 2, centreName: "Centre 2") };
+            A.CallTo(() => userService.GetUsersByUsername(A<string>._))
+                .Returns((expectedAdminUser, expectedDelegateUsers));
+            A.CallTo(() => loginService.VerifyUsers(A<string>._, A<AdminUser>._, A<List<DelegateUser>>._))
+                .Returns((expectedAdminUser, expectedDelegateUsers));
+
+            // When
+            var result =
+                controller.Index(LoginTestHelper.GetDefaultLoginViewModel());
+
+            // Then
+            result.Should().BeViewResult().WithViewName("ChooseACentre");
+        }
+
+        [Test]
+        public void Multiple_accounts_with_different_passwords_should_use_ones_with_input_password()
+        {
+            // Given
+            controller = LoginTestHelper.GetLoginControllerWithSignInFunctionality(loginService, userService);
+            var expectedAdminUser = UserTestHelper.GetDefaultAdminUser(centreId: 1, centreName: "Centre 1");
+            var expectedDelegateUsers = new List<DelegateUser>
+                { UserTestHelper.GetDefaultDelegateUser(centreId: 2, centreName: "Centre 2") };
+            A.CallTo(() => userService.GetUsersByUsername(A<string>._))
+                .Returns((expectedAdminUser, expectedDelegateUsers));
+            A.CallTo(() => loginService.VerifyUsers(A<string>._, A<AdminUser>._, A<List<DelegateUser>>._))
+                .Returns((expectedAdminUser, new List<DelegateUser>()));
+
+            // When
+            var result =
+                controller.Index(LoginTestHelper.GetDefaultLoginViewModel());
+
+            // Then
+            result.Should().BeRedirectToActionResult()
+                .WithControllerName("Home").WithActionName("Index");
+        }
+
+        [Test]
+        public void Multiple_accounts_with_different_approved_statuses_should_use_accounts_with_approved_statuses()
```

Comment:
This is good, but the test name is a bit misleading - this test would pass if it only used accounts with unapproved statuses (same with the password test above).

Reply:
So should the test itself be kept the same, but rename it to something like "When Log In finds approved and unapproved accounts it should log in to approved accounts only"?
And the password test becomes "When Log In finds accounts with different passwords it should log in to accounts that match input password only"

Reply:
Maybe something like:
"When_user_has_multiple_accounts_with_different_passwords_redirects_home_rather_than_choose_a_centre" (or redirects home/does not redirect to choose a centre)
I think the main thing is that we don't test which user is logged in, but that both are not considered valid accounts for Choose a centre page. So "only_log_in_to_ones_matching_input_password" could not be true (e.g. we could only log in the account where password doesn't match), but the test would pass .

Reply:
When initially writing the test, I was trying to figure out the best way to check that accounts without password matches were not available, and could only arbitrarily do this by making the case where:
- If all accounts were seen as acceptable for logging in, then there would be too many centres and thus would be redirected to the Choose A Centre page
- If accounts without password matches were excluded, then the acceptable accounts only have 1 centre, and thus log in and get taken to the Home page

I think I see now what you mean, that the test itself is not specific enough. However, I'm not sure your suggested test names make the situation any better - that name would not work if the user had 5 different delegate accounts at different centres, and some of them had the correct password and some did not.

I think the test itself needs redoing, as the point was to make sure that accounts that don't match the password are not considered acceptable. Currently, I don't think there is an easy way to do this, but your other comment about checking available centres in a helper/service might be the way forward, so I'll refactor into using that.